### PR TITLE
Initial Implementation of PMM Curve

### PR DESCRIPTION
**Pull request by KhaledEmaraDev**

### CODE REVIEW

**Comment by joncinque** [comment]
Watch out, this pulled in a lot of other changes!

File: token-swap/program/src/curve/pmm.rs [hunk 1]
```diff
@@ -0,0 +1,906 @@
...
+                        PreciseNumber::new(amount)?
+                            .checked_sub(&PreciseNumber::new(back_to_one_receive_base)?)?
+                            .to_imprecise()?,
+                        new_base_target,
+                    )?)?)?
+                    .to_imprecise()?,
+            );
+        }
+    } else {
+        Some(r_below_sell_base_token(
```
Comment (joncinque) [inline token-swap/program/src/curve/pmm.rs:649]: should be `_RAboveBuyBaseToken`

**Review by joncinque — COMMENTED** [review]
Wow, great job on all this! It's tricky to keep all of these functions in line and all of this math, but you really fought through it.  A few comments on how we can make this better and some parts where I noticed deviations from the solidity implementation.  Let me know if you have any questions

**Comment by KhaledEmaraDev** [comment]
Thanks @joncinque for the review. I'll work on these during this week and request your re-review.

**Comment by jnappy** [comment]
@KhaledEmaraDev bumping this - any more progress on this PR?

**Comment by KhaledEmaraDev** [comment]
@jnappy I have pushed @joncinque requested changes.

**Review by joncinque — COMMENTED** [review]
The changes look good! Next you'll need to run cargo fmt and cargo clippy to fixup some little issues and get past CI.  Also, it seems like some of the new tests are failing

**Pull request closed without merging**